### ISSUE TITLE
Bug: `display: none` on react component with linearGradient svg also affects svgs outside component

### ISSUE THREAD

**Issue by kimbaudi** [facebook/react#21064]
I have a svg react component with linearGradient that I render in my Header component and my main component. When I set `display: none` style to the Header component, I expect the svg in the Header component to be hidden and the svg in the main component to be visible. When I use svg without linearGradient, I get the expected behavior of svg hidden in Header and visible in main component. However, if the svg uses linearGradient, then setting `display: none` to the Header **hides the svg in the Header and the main component**. 

To reproduce the issue, I have 2 react components `PythonIcon` (using linearGradient) and `WebpackIcon` (not using linearGradient). In my `App.js`, I render the Header and also render both `PythonIcon` and `WebpackIcon` in my main component:

```jsx
import React from "react";
import Header from "./Header";
import PythonIcon from "./PythonIcon";
import WebpackIcon from "./WebpackIcon";

function App() {
  return (
    <div>
      <Header />
      <main style={{ marginTop: "1rem", border: "1px solid green" }}>
        <h1>
          <PythonIcon width="20" height="20" />
          Python
        </h1>
        <h1>
          <WebpackIcon width="20" height="20" />
          Webpack
        </h1>
      </main>
    </div>
  );
}

export default App;
```

In my Header component, I also render `PythonIcon` and `WebpackIcon`

```jsx
import React from "react";
import PythonIcon from "./PythonIcon";
import WebpackIcon from "./WebpackIcon";

export default function Header() {
  return (
    <header style={{ border: "1px solid red", display: "block" }}>
      <h1>
        <PythonIcon width="20" height="20" />
        Python
      </h1>
      <h1>
        <WebpackIcon width="20" height="20" />
        Webpack
      </h1>
    </header>
  );
}
```

And the page looks like this:

![a](https://user-images.githubusercontent.com/1864092/112250537-c6f44b00-8c16-11eb-83ac-12fb554966da.png)

But once I change `display: block` to `display: none` in the Header component, the page looks like this:

![b](https://user-images.githubusercontent.com/1864092/112250604-e68b7380-8c16-11eb-8bd2-ff503ace0822.png)

Notice that the webpack icon (which doesn't use linearGradient) is correctly showing in main component, but the python icon (which uses linearGradient) is not. 

This was unexpected. Am I doing something wrong? Is there any way to hide the python icon in the header without having it hide the python icon in the main component? I am trying to create a responsive navbar using css media-query to show/hide navbar. Unfortunately, hiding the navbar w/ svg icons also hides the svg icon in my other components that I render.

React version: 17.0.2

## Steps To Reproduce

1. clone `https://github.com/kimbaudi/react-svg-lineargradient-issue.git`
2. `npm install && npm start` (or yarn)
3. change between `display: block` and `display: none` in Header component

Link to code example:

I've provided a minimal reproduction at https://github.com/kimbaudi/react-svg-lineargradient-issue

## The current behavior

hiding Header containing svg w/ linearGradient should only hide svg inside the Header, but not outside the Header.

## The expected behavior

hiding Header containing svg w/ linearGradient also hides svg rendered outside of Header.

**Comment by kimbaudi** [comment]
I think I have an idea why I am getting this issue. It is probably because using linearGradient in svgs require `id` so rendering multiple svgs using linearGradient will render multiple elements with the same id. I'll need to see if passing unique id as prop to the svgs w/ linearGradient solves the issue before I can safely close this.

**Comment by eps1lon** [comment]
I would also suggest trying this with just HTML+CSS and see if this behaves the same. Comparing HTML and React would also help maintainers identify the issue.

**Comment by Andarist** [comment]
@kimbaudi this is exactly the case here - IDs need to be unique: https://codesandbox.io/s/vigorous-lake-c4yvt?file=/src/index.js:3041-3055

**Comment by kimbaudi** [comment]
@Andarist thank you for the confirmation. My issue is not related to react, but actually related to svgs containing `id`s as is the case with the PythonIcon svg using linearGradient. One thing I noticed from your codesandbox is that the Python svg icon loses the linearGradient and just shows as black:

![c](https://user-images.githubusercontent.com/1864092/112327383-9e9a3a00-8c72-11eb-9584-d8aff84b3992.png)

So basically **avoid svgs using linearGradient as icons**. 

closing this issue as it is not related to react, but expected result of using svgs containing ids in multiple components.

**Comment by Andarist** [comment]
I've copy-pasted your code to create a plain HTML demo and I've forgotten to replace `stopColor` with `stop-color` in it. It works just OK when I fix it: https://codesandbox.io/s/exciting-thompson-qgwj3?file=/src/index.js:3045-3059 .

So don't be afraid of using `linearGradient`! :wink:

**Comment by kimbaudi** [comment]
So I found a solution where I am still able to use svg react components containing `id`s (linearGradient, radialGradient). Basically, I append a randomly generated string to each id. Now hiding the python icon in Header component doesn't hide the python icon in the main component: 

![d](https://user-images.githubusercontent.com/1864092/112334885-03589300-8c79-11eb-93dd-ea5a12a5316f.png)

So here is my previous PythonIcon react svg component that was causing issue due to hardcoded id:

```jsx
import React from "react";

const PythonIcon = (props) => (
  <svg
    xmlns="http://www.w3.org/2000/svg"
    viewBox="0 0 64 64"
    width="64"
    height="64"
    {...props}
  >
    <linearGradient
      id="pythonLinearGradient1"
      x1="611.11"
      x2="720.13"
      y1="-280.51"
      y2="-374.3"
      gradientTransform="matrix(.32049 0 0 -.32362 -197.36 -88.951)"
      gradientUnits="userSpaceOnUse"
    >
      <stop stopColor="#5A9FD4" offset="0" />
      <stop stopColor="#306998" offset="1" />
    </linearGradient>
    <path
      fill="url(#pythonLinearGradient1)"
      strokeWidth=".56976"
      d="m31.344 0c-2.6209 0-5.1278 0.2279-7.2929 0.62674-6.4383 1.1395-7.6348 3.5325-7.6348 7.9197v5.8116h15.27v1.9372h-21.024c-4.4441 0-8.3185 2.6779-9.572 7.7488-1.4244 5.8116-1.4814 9.458 0 15.498 1.0825 4.5011 3.7034 7.7488 8.1476 7.7488h5.2418v-6.9511c0-5.0139 4.3872-9.515 9.572-9.515h15.27c4.2732 0 7.6348-3.4755 7.6348-7.7488v-14.529c0-4.1593-3.4755-7.236-7.6348-7.9197-2.6209-0.45581-5.3558-0.62674-7.9767-0.62674zm-8.2615 4.672c1.5953 0 2.8488 1.3104 2.8488 2.9058 0 1.5953-1.3104 2.9058-2.8488 2.9058-1.5953 0-2.8488-1.3104-2.8488-2.9058 0-1.5953 1.2535-2.9058 2.8488-2.9058z"
    />
    <linearGradient
      id="pythonLinearGradient2"
      x1="762.27"
      x2="723.34"
      y1="-431.07"
      y2="-375.99"
      gradientTransform="matrix(.32049 0 0 -.32362 -197.36 -88.951)"
      gradientUnits="userSpaceOnUse"
    >
      <stop stopColor="#FFD43B" offset="0" />
      <stop stopColor="#FFE873" offset="1" />
    </linearGradient>
    <path
      fill="url(#pythonLinearGradient2)"
      strokeWidth=".56976"
      d="m48.836 16.352v6.7802c0 5.2418-4.4441 9.6859-9.572 9.6859h-15.213c-4.1593 0-7.6348 3.5895-7.6348 7.7488v14.529c0 4.1593 3.5895 6.5523 7.6348 7.7488 4.843 1.4244 9.458 1.6523 15.27 0 3.8744-1.1395 7.6348-3.3616 7.6348-7.7488v-5.8116h-15.27v-1.9372h22.904c4.4441 0 6.0964-3.0767 7.6348-7.7488 1.5953-4.786 1.5384-9.4011 0-15.498-1.0825-4.4441-3.1907-7.7488-7.6348-7.7488zm-8.5464 36.807c1.5953 0 2.8488 1.3104 2.8488 2.9058 0 1.5953-1.3104 2.9058-2.8488 2.9058-1.5953 0-2.8488-1.3105-2.8488-2.9058-0.05698-1.5953 1.2535-2.9058 2.8488-2.9058z"
    />
  </svg>
);

export default PythonIcon;
```

and here is the updated PythonIcon react svg component with unique ids:

```jsx
import React from "react";

const PythonIcon = (props) => {
  const rand1 = Math.random()
  const rand2 = Math.random()
  return (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      viewBox="0 0 64 64"
      width="64"
      height="64"
      {...props}
    >
      <linearGradient
        id={`pythonLinearGradient1-${rand1}`}
        x1="611.11"
        x2="720.13"
        y1="-280.51"
        y2="-374.3"
        gradientTransform="matrix(.32049 0 0 -.32362 -197.36 -88.951)"
        gradientUnits="userSpaceOnUse"
      >
        <stop stopColor="#5A9FD4" offset="0" />
        <stop stopColor="#306998" offset="1" />
      </linearGradient>
      <path
        fill={`url(#pythonLinearGradient1-${rand1})`}
        strokeWidth=".56976"
        d="m31.344 0c-2.6209 0-5.1278 0.2279-7.2929 0.62674-6.4383 1.1395-7.6348 3.5325-7.6348 7.9197v5.8116h15.27v1.9372h-21.024c-4.4441 0-8.3185 2.6779-9.572 7.7488-1.4244 5.8116-1.4814 9.458 0 15.498 1.0825 4.5011 3.7034 7.7488 8.1476 7.7488h5.2418v-6.9511c0-5.0139 4.3872-9.515 9.572-9.515h15.27c4.2732 0 7.6348-3.4755 7.6348-7.7488v-14.529c0-4.1593-3.4755-7.236-7.6348-7.9197-2.6209-0.45581-5.3558-0.62674-7.9767-0.62674zm-8.2615 4.672c1.5953 0 2.8488 1.3104 2.8488 2.9058 0 1.5953-1.3104 2.9058-2.8488 2.9058-1.5953 0-2.8488-1.3104-2.8488-2.9058 0-1.5953 1.2535-2.9058 2.8488-2.9058z"
      />
      <linearGradient
        id={`pythonLinearGradient2-${rand2}`}
        x1="762.27"
        x2="723.34"
        y1="-431.07"
        y2="-375.99"
        gradientTransform="matrix(.32049 0 0 -.32362 -197.36 -88.951)"
        gradientUnits="userSpaceOnUse"
      >
        <stop stopColor="#FFD43B" offset="0" />
        <stop stopColor="#FFE873" offset="1" />
      </linearGradient>
      <path
        fill={`url(#pythonLinearGradient2-${rand2})`}
        strokeWidth=".56976"
        d="m48.836 16.352v6.7802c0 5.2418-4.4441 9.6859-9.572 9.6859h-15.213c-4.1593 0-7.6348 3.5895-7.6348 7.7488v14.529c0 4.1593 3.5895 6.5523 7.6348 7.7488 4.843 1.4244 9.458 1.6523 15.27 0 3.8744-1.1395 7.6348-3.3616 7.6348-7.7488v-5.8116h-15.27v-1.9372h22.904c4.4441 0 6.0964-3.0767 7.6348-7.7488 1.5953-4.786 1.5384-9.4011 0-15.498-1.0825-4.4441-3.1907-7.7488-7.6348-7.7488zm-8.5464 36.807c1.5953 0 2.8488 1.3104 2.8488 2.9058 0 1.5953-1.3104 2.9058-2.8488 2.9058-1.5953 0-2.8488-1.3105-2.8488-2.9058-0.05698-1.5953 1.2535-2.9058 2.8488-2.9058z"
      />
    </svg>
  );
};

export default PythonIcon;
```

**Comment by kimbaudi** [comment]
@Andarist - thank you so much. I just realized that I can still use linearGradient & radialGradient in my svg react components. I've made the change on my minimal repo: https://github.com/kimbaudi/react-svg-lineargradient-issue

**Comment by vclau-1996** [comment]
I think you can use the img tag to show the svg that has lineartGradient. 

like below codes:

```typescript
import pythonIconPath from '../img/pythonIcon.svg';

const PythonIcon=()=>{
    return <img src={pythonIconPath} alt='' />;
};

export default PythonIcon;

```

you can try the method.
